de_project/de_app/models.py

```diff
@@ -1,3 +1,206 @@
+class Metric(models.Model):
+    name = models.CharField(default="INVALID_METRIC", max_length=255)
+    current_value = models.IntegerField(default=0)
+    total_value = models.IntegerField(default=0)
+    average_value = models.IntegerField(default=0)
+
+    def __str__(self):
+        return self.name
+
+    def validate(self):
+        #Other validation can be done here
+        self.save()
+
+    def reset(self):
+        self.current_value = 0
+        self.total_value = 0
+        self.average_value = 0
+
+    def get_value_names(self):
+        return [ "current_value", "total_value", "average_value" ]
+
+    def set_value(self, value_name, value):
+        #TODO: make this dynamic? we might want more calculated metric values...
+        if value_name == "current_value":
+            self.current_value = value
+        elif value_name == "total_value":
+            self.total_value = value
+        elif value_name == "average_value":
+            self.average_value = value
+        else:
+            raise ValueError("No such value {value_name} in Metric {self.name}")
+        self.validate()
+
+    def get_value(self, value_name):
+        #TODO: make this dynamic? we might want more calculated metric values...
+        if value_name == "current_value":
+            return self.current_value
+        elif value_name == "total_value":
+            return self.total_value
+        elif value_name == "average_value":
+            return self.average_value
+        else:
+            raise ValueError("No such value {value_name} in Metric {self.name}")
+
+    def get_metric(metric_name, metrics):
+        for m in metrics:
+            if m.name == metric_name:
+                return m
+        raise ValueError("No Metric {metric_name} in {metrics}")
+
+class MetricSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = Metric
+        fields = ["name", "current_value", "total_value", "average_value"]
+
+
+class Record(models.Model):
+    timestamp = models.DateTimeField(auto_now_add=True, blank=True)
+    location = models.CharField(default="", max_length=255)
+    metrics = models.ManyToManyField(Metric)
+
+    def validate(self):
+        #Other validation can be done here
+        self.save()
+
+    def create_metrics(self, metric_names_to_create = []):
+        for n in metric_names_to_create:
+            toAdd = Metric(name=n)
+            toAdd.validate()
+            self.metrics.add(toAdd)
+
+    def set_metric_value(self, metric_name, metric_value, value):
+        self.metrics.get(name=metric_name).set_value(metric_value, value)
+
+    def get_metric_value(self, metric_name, metric_value):
+        return self.metrics.get(name=metric_name).get_value(metric_value)
+
+    def get_metric(self, metric_name):
+        return Metric.get_metric(metric_name, self.get_all_metrics())
+
+    def get_all_metrics(self):
+        return self.metrics.all()
+
+    def do_metrics_contain(self, metrics):
```

Comment:
I still think this method name is extremely confusing

Reply:
I'll do a survey and see what other people prefer.